src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpTypingStyleDiagnosticAnalyzerBase.State.cs

```diff
@@ -0,0 +1,237 @@
+                    return false;
+                }
+
+                var methodSymbol = semanticModel.GetSymbolInfo(memberName, cancellationToken).Symbol as IMethodSymbol;
+                if (methodSymbol == null)
+                {
+                    return false;
+                }
+
+                if (memberName.IsRightSideOfDot())
+                {
+                    var typeName = memberName.GetLeftSideOfDot();
+                    return IsPossibleCreationOrConversionMethod(methodSymbol, declaredTypeSymbol, semanticModel, typeName, cancellationToken);
+                }
+
+                return false;
+            }
+
+            private bool IsIntrinsicType(SyntaxNode declarationStatement) =>
+                declarationStatement.IsKind(SyntaxKind.VariableDeclaration)
+                ? ((VariableDeclarationSyntax)declarationStatement).Variables.Single().Initializer.Value.IsAnyLiteralExpression()
+                : false;
+
+            private ExpressionSyntax GetInitializerExpression(EqualsValueClauseSyntax initializer) =>
+                initializer.Value is CheckedExpressionSyntax
+                    ? ((CheckedExpressionSyntax)initializer.Value).Expression
+                    : initializer.Value;
+
+            private bool IsPossibleCreationOrConversionMethod(IMethodSymbol methodSymbol, ITypeSymbol declaredType, SemanticModel semanticModel, ExpressionSyntax typeName, CancellationToken cancellationToken)
+            {
+                if (methodSymbol.ReturnsVoid)
+                {
+                    return false;
+                }
+
+                var typeInInvocation = semanticModel.GetTypeInfo(typeName, cancellationToken).Type;
+
+                return IsPossibleCreationMethod(methodSymbol, declaredType, typeInInvocation)
+                    || IsPossibleConversionMethod(methodSymbol, declaredType, typeInInvocation, semanticModel, cancellationToken);
+            }
+
+            private bool IsPossibleCreationMethod(IMethodSymbol methodSymbol, ITypeSymbol declaredType, ITypeSymbol typeInInvocation)
+            {
+                var isTypeInfoSame = false;
+
+                // Pattern: Method name is a prefix match of one of these well known method names
+                //          and method is a member of type being created.
+                // cases: int.ParseXXX, TextSpan.FromBounds, 
+                if (methodSymbol.Name.StartsWith("Parse", StringComparison.Ordinal)
+                    || methodSymbol.Name.StartsWith("From", StringComparison.Ordinal))
+                {
+                    isTypeInfoSame = typeInInvocation.Equals(declaredType);
+                }
+
+                // Pattern: Method name is an exact match of one of these well known creation methods
+                //          and method is a member of type being created.
+                // cases: node.With, Tuple.Create, State.Generate
+                if (methodSymbol.Name.Equals("With", StringComparison.Ordinal)
```

Comment:
Also note: you may end up wanting to match type names instead of using type equality.  that will ensure that if the user calls "Tuple.Create(0, true)" that we'll consider that sufficient.  This is because Tuple.Create returns a ```Tuple<X,Y>``` which is not the same type as ```Tuple```.